src/disc_bsd.c

```diff
@@ -23,29 +23,37 @@
 
 --------------------------------------------------------------------------- */
 
+#include <assert.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <string.h>
-#include <sys/types.h>
 #include <sys/cdio.h>
 #include <sys/ioctl.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#if !defined(__FreeBSD__)
+#include <util.h>
+#endif
 
 #include "discid/discid_private.h"
 #include "unix.h"
 
-int mb_disc_unix_read_toc_header(int fd, mb_disc_toc *toc) {
-	struct ioc_toc_header th;
-
-	int ret = ioctl(fd, CDIOREADTOCHEADER, &th);
+#define MAX_DEV_LEN 15
 
-	if ( ret < 0 )
-		return 0; /* error */
-
-	toc->first_track_num = th.starting_track;
-	toc->last_track_num = th.ending_track;
-
-	return 1;
+static int get_device(int n, char* device_name, size_t device_name_length) {
+#if !defined(__FreeBSD__) /* /dev/rcdNX, where X is the letter for the raw partition */
+	snprintf(device_name, device_name_length, "/dev/rcd%d%c", n - 1, 'a' + getrawpartition());
+#else /* on FreeBSD it's just /dev/cdN */
+	snprintf(device_name, device_name_length, "/dev/cd%d", n - 1);
+#endif
+	return mb_disc_unix_exists(device_name);
 }
 
+/* not actually used for *BSD, but needed by unix.c */
```

Comment:
I don't quite get why you define these functions as no-op and put the code directly into mb_disc_read_unportable.

mb_disc_read_unportable() is a mess duplicating especially error checking code from unix.c and code blocks that could happily life inside of the functions that are now empty in disc_bsd.c

I don't get why you don't want to use mb_disc_unix_read  (like disc_linux.c does)  or mb_disc_unix_read_toc